### DISCUSSION TITLE
How to use the `gguf-split` / Model sharding demo

### DISCUSSION THREAD

**Discussion by phymbert** [ggml-org/llama.cpp#6404] [Show and tell]
#### Context
Distributing and storing GGUFs is difficult for 70b+ models, especially on f16. Lot of issue can happen during file transfers, examples:

- temporary disk full
- network interruption

Typically, GGUFs need to be transferred from [Hugging Face](https://huggingface.co/ggml-org) to an internal storage like [s3](https://aws.amazon.com/s3/?nc1=h_ls), [minio](https://min.io/), [git lfs](https://github.com/git-lfs/git-lfs), [nexus](https://www.sonatype.com/products/sonatype-nexus-repository) or [artifactory](https://jfrog.com/artifactory/), then downloaded by the inference server and stored locally (or on a k8s PvC for example). 

Storage solutions and filesystems poorly support large GGUF, typically HF does not support files larger than [50GB](https://huggingface.co/docs/hub/repositories-recommendations). 
Such limits also exist on [Artifactory](https://jfrog.com/help/r/does-artifactory-limit-the-file-upload-size-when-uploading-using-the-rest-api).

#### Solution

We recently introduced `gguf-split` CLI and support the load of sharded GGUFs model in `llama.cpp`:

- #6135
- #6187
- #6192
- #6234
- #6343

##### Download a model

```python
from huggingface_hub import snapshot_download
snapshot_download(repo_id="keyfan/grok-1-hf")
```

##### Convert to GGUF F16

```shell
python -u convert-hf-to-gguf.py \
  ~/.cache/huggingface/hub/models--keyfan--grok-1-hf/snapshots/64e7373053c1bc7994ce427827b78ec11c181b3e/ \
  --outfile grok-1-f16.gguf \
  --outtype f16
```

NOTE: Follow llama.cpp build instructions to generate all tools/cli: `make`.

##### Quantize (optional)
```shell
quantize grok-1-f16.gguf grok-1-q4_0.gguf q4_0
```

#### Build model shards
It is possible to use different sharding strategy:
- Max tensors per file: `--split-max-tensors 256`
- Max file size: `--split-max-size 48G`

```shell
gguf-split --split --split-max-tensors 256 grok-1-q4_0.gguf grok-1-q4_0
```

It will produce 9 files with maximum 256 tensors in each.

You can then [upload the sharded model to your HF Repo](https://huggingface.co/docs/huggingface_hub/guides/upload):

```shell
huggingface-cli upload [repo_id] [local_path] [path_in_repo]
```

Files produced by `gguf-split` are valid GGUFs, so you can [visualize them in HF](https://huggingface.co/ggml-org/models?show_tensors=grok-1%2Fgrok-1-q4_0-00001-of-00009.gguf).

##### Load sharded model

`llama_load_model_from_file` will detect the number of files and will load additional tensors from the rest of files.
```shell
main --model grok-1-q4_0-00001-of-00009.gguf -ngl 64
```
You may notice:
> llama_model_loader: additional 8 GGUFs metadata loaded.

##### Load sharded model from a remote URL
```shell
main --hf-repo ggml-org/models \
  --hf-file grok-1/grok-1-q4_0-00001-of-00009.gguf \
  --model   models/grok-1-q4_0-00001-of-00009.gguf \
  -ngl 64
```

**Comment by maziyarpanahi** [comment]
Maybe I am wrong, but I couldn't make `--split-max-size 48G` work.

**Reply by christianazinn** [reply]
Nor I. Splitting an imatrix quant of Mixtral 8x22b with `split-max-size 20G` tries to write three ~1G tensors, one 51G tensor, and one 22G tensor. Strange.

**Reply by phymbert** [reply]
Yes it's bugged @ngxson

**Reply by phymbert** [reply]
@christianazinn Use the `--split-max-tensors N`  meanwhile

**Reply by ngxson** [reply]
@christianazinn Can you please run `--split-max-size 48G --dry-run` and post the result here? The "dry run" option will calculate the split plan, without writing new files

@phymbert I've never tested with models bigger than 5GB. The workflow test (that I'll add when I come back from vacation) only uses the smaller test model. Do you have any idea if we can test big models?

**Reply by phymbert** [reply]
~I do not think the size of the model matter here. Just need to put a size which is coherent with the size of each tensor. Once we have identified the issue, we can reproduce with tinnyllama.~ 
Enjoy your vacations bro, there is no emergency at all here :)

EDIT: https://github.com/ggerganov/llama.cpp/discussions/6404#discussioncomment-9083158

**Reply by maziyarpanahi** [reply]
`--split-max-tensors N` worked perfectly for 16bit, for the rest I went without any flag and it did 5 splits by default. Thank you for this feature or else this would have been an actual hell: https://huggingface.co/MaziyarPanahi/Mixtral-8x22B-v0.1-GGUF

**Reply by slaren** [reply]
There is an integer overflow in `split_str_to_n_bytes` that prevents using splits larger than 2GB.

**Reply by christianazinn** [reply]
@ngxson Here it is. 
![Screenshot 2024-04-11 080953](https://github.com/ggerganov/llama.cpp/assets/59622928/0097d81d-b78b-40e8-a4ee-416bbb2d8232)

The really strange part is that it looks like changing the parameter very slightly results in drastically different outputs:
![Screenshot 2024-04-11 081135](https://github.com/ggerganov/llama.cpp/assets/59622928/496a36db-e44f-411b-9c7f-87e1198608cc)

**Reply by christianazinn** [reply]
Ah, and I'm not so lucky as @maziyarpanahi with splitting Mixtral 8x22b. Only possible difference is that I quantized each model with an imatrix, but I'm not getting nicely sized splits sans flags.

**Reply by 4cecoder** [reply]
please for the love of gpt how do you merge gguf files? 

there isn't much docs on the net? 

I tried compiling llama.cpp on windows and i've got that but now what?

Just want to merge 1-5 sharded gguf of command r+!

**Reply by christianazinn** [reply]
@4cecoder `gguf-split --merge [path-to-first-shard] [path-to-outfile]` should do the trick.

**Reply by 4cecoder** [reply]
I don't have that `gguf-split` command by itself 

is it its own binary? Where's the source to compile it? 

**wait found gguf-split but it will not compile...**



```bash
PS D:\Repos\llama.cpp\gguf-split> g++.exe -I D:\Repos\llama.cpp -I D:\Repos\llama.cpp\common -o gguf-split gguf-split.cpp ../ggml.c ../llama.cpp
```

**Comment by dranger003** [comment]
On Windows you can compile llama.cpp by opening a VS native tools command prompt (i.e. `x64 Native Tools Command Prompt for VS 2022`) and running the commands below. Make sure your VS install has C++ dev libraries installed.
```
git clone --recursive https://github.com/ggerganov/llama.cpp
cmake -S . -B build && cmake --build build --config Release
build\bin\Release\gguf-split.exe -h
```

**Comment by taozhiyuai** [comment]
@dranger003 @phymbert may I ask how to compile gguf-split on MAC?

(llamacpp) taozhiyu@603e5f4a42f1 llama.cpp-master % gguf-split
zsh: command not found: gguf-split

**Reply by phymbert** [reply]
Hum, I am Mac less. @ggerganov, could you please help ?

Is not simply calling `make` not working ?

**Reply by ggerganov** [reply]
```bash
make
./gguf-split
```

**Reply by phymbert** [reply]
Summary updated ^)

**Reply by taozhiyuai** [reply]
Works on Mac.thanks.

'make
./gguf-split'

**Reply by matbeedotcom** [reply]
I dont have to compile this whole project just to merge gguf's do i?

**Comment by bennmann** [comment]
Please also include more clear and specific instructions for --merge?

**Reply by phymbert** [reply]
Merging is not necessary anymore in most use cases.

**Reply by 4cecoder** [reply]
Merging in LLM studio is no longer necessary so just use LLM studio on windows @bennmann

**Reply by q5sys** [reply]
@phymbert "most cases" is not all cases, and it would be nice if this was clearly documented with full example for those cases that fall outside of "most".  Front ends that use llama.cpp as a backend are sometimes not coded to operate with split files. and so people using those need to merge themselves.   I dont know if it still is, but oobabooga is an example (at least a few weeks ago when I last tested it) where a UI was unable to deal with split files.  

People are still running into issues trying to use this tool.  As an example here's two in the past week. https://huggingface.co/leafspark/Meta-Llama-3.1-405B-Instruct-GGUF/discussions/2
